lib/fluent/plugin/file_wrapper.rb

```diff
@@ -16,8 +16,34 @@
 
 module Fluent
   module FileWrapper
-    def self.open(*args)
-      io = WindowsFile.new(*args).io
+    include File::Constants
+    def self.mode2flags(mode)
+      # Always need BINARY to enable SHARE_DELETE
+      # https://bugs.ruby-lang.org/issues/11218
+      # https://github.com/ruby/ruby/blob/d6684f063bc53e3cab025bd39526eca3b480b5e7/win32/win32.c#L6332-L6345
+      flags = BINARY | SHARE_DELETE
+      case mode.delete("b")
+      when "r"
+        flags |= RDONLY
+      when "r+"
+        flags |= RDWR
+      when "w"
+        flags |= WRONLY | CREAT | TRUNC
+      when "w+"
+        flags |= RDWR | CREAT | TRUNC
+      when "a"
+        flags |= WRONLY | CREAT | APPEND
+      when "a+"
+        flags |= RDWR | CREAT | APPEND
+      else
+        raise Errno::EINVAL.new("Unsupported mode by Fluent::FileWrapper: #{mode}")
+      end
+    end
+
+    def self.open(path, mode='r')
+      # inject File::Constants::SHARE_DELETE
+      # https://github.com/fluent/fluentd/pull/3585#issuecomment-1101502617
+      io = File.open(path, mode2flags(mode))
```

Comment:
My Win32-fu is somewhat rusty, but won't this make Fluentd acquire exclusive
read/write lock on the target file? I'm thinking a scenario like this:

1. Fluentd started watching on `access.log`, and while doing so,
    it acquired read/write lock on the file.
2. Nginx started and it tried to open `access log` to record requests.
3. ???

To avoid the issue above, we probably need to set  `FILE_SHARE_READ` and
`FILE_SHARE_WRITE` flags (or will Ruby set these flags by default? I haven't
fully trace the file opening code in Ruby 3.x...)

Reply:
> To avoid the issue above, we probably need to set `FILE_SHARE_READ` and `FILE_SHARE_WRITE` flags (or will Ruby set these flags by default? I haven't fully trace the file opening code in Ruby 3.x...)

They are always added:
https://github.com/ruby/ruby/blob/c7bbed299493d0fe5323916b83dab04021bc570f/win32/win32.c#L6446
```c
	h = CreateFileW(file, access, FILE_SHARE_READ | FILE_SHARE_WRITE | share_delete, &sec, create, attr, NULL);
```

Reply:
OK! LGTM.